File: frontend/src/types/incident.ts
import { User } from './user'

export type IncidentSeverity = 'low' | 'medium' | 'high' | 'critical'
export type IncidentStatus = 'open' | 'in_progress' | 'resolved' | 'closed'

export interface Incident {
  id: number
  reported_by_id: number
  assigned_to_id?: number | null
  title: string
  description: string
  severity: IncidentSeverity
  status: IncidentStatus
  resolution_notes?: string | null
  created_at: string
  updated_at: string
  resolved_at?: string | null
  reported_by?: User | null
  assigned_to?: User | null
}

export interface IncidentCreate {
  title: string
  description: string
  severity?: IncidentSeverity
  assigned_to_id?: number | null
}

export interface IncidentUpdate {
  title?: string
  description?: string
  severity?: IncidentSeverity
  assigned_to_id?: number | null
}

export interface IncidentStatusUpdate {
  status: IncidentStatus
  resolution_notes?: string | null
}

export interface IncidentAssign {
  assigned_to_id?: number | null
}

export interface IncidentList {
  items: Incident[]
  total: number
  page: number
  limit: number
}

File: backend/app/schemas/incident.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, Literal, List
from app.schemas.user import UserResponse


class IncidentBase(BaseModel):
    title: str = Field(..., max_length=200)
    description: str = Field(..., max_length=5000)
    severity: Literal["low", "medium", "high", "critical"] = "medium"
    assigned_to_id: Optional[int] = None


class IncidentCreate(IncidentBase):
    pass


class IncidentUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = Field(None, max_length=5000)
    severity: Optional[Literal["low", "medium", "high", "critical"]] = None
    assigned_to_id: Optional[int] = None


class IncidentStatusUpdate(BaseModel):
    status: Literal["open", "in_progress", "resolved", "closed"]
    resolution_notes: Optional[str] = Field(None, max_length=5000)


class IncidentAssign(BaseModel):
    assigned_to_id: Optional[int] = None


class Incident(IncidentBase):
    id: int
    reported_by_id: int
    status: Literal["open", "in_progress", "resolved", "closed"]
    resolution_notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    resolved_at: Optional[datetime] = None
    reported_by: Optional[UserResponse] = None
    assigned_to: Optional[UserResponse] = None
    
    class Config:
        from_attributes = True


class IncidentList(BaseModel):
    items: List[Incident]
    total: int
    page: int
    limit: int
